test/Kestrel.Core.Tests/Http2ConnectionTests.cs

```diff
@@ -2886,6 +2954,158 @@ public async Task ConnectionResetNotLoggedWithNoActiveStreams()
+        }
+
+        [Fact]
+        public async Task CompletionSendsFinalGOAWAY()
+        {
+            await InitializeConnectionAsync(_noopApplication);
+
+            // Completes ProcessRequestsAsync
+            _pair.Application.Output.Complete();
+            Assert.True(_connection.StateTransitionEvents[Http2Connection.ConnectionState.Closed].Wait(5000));
+
+            VerifyGoAway(await ReceiveFrameAsync(), 0, Http2ErrorCode.NO_ERROR);
+        }
+
+        [Fact]
+        public async Task GracefulGOAWAYSentByStopProcessingNextRequest()
+        {
+            await InitializeConnectionAsync(_waitForAbortApplication);
+
+            await StartStreamAsync(1, _browserRequestHeaders, endStream: true);
+
+            Assert.Equal(Http2Connection.ConnectionState.Open, _connection.State);
+            Assert.Equal(1, _connection.HighestOpenedStreamId);
+
+            _connection.StopProcessingNextRequest();
+            Assert.True(_connection.StateTransitionEvents[Http2Connection.ConnectionState.Closing].Wait(5000));
+
+            VerifyGoAway(await ReceiveFrameAsync(), Int32.MaxValue, Http2ErrorCode.NO_ERROR);
+
+            _connection.OnInputOrOutputCompleted();
+            Assert.True(_connection.StateTransitionEvents[Http2Connection.ConnectionState.Closed].Wait(5000));
+
+            VerifyGoAway(await ReceiveFrameAsync(), 1, Http2ErrorCode.NO_ERROR);
+
+            await SendRstStreamAsync(1);
+            await WaitForAllStreamsAsync();
+        }
+
+        [Fact]
+        public async Task FinalGOAWAYSentByStopProcessingNextRequestWhenAllStreamsComplete()
+        {
+            await InitializeConnectionAsync(_waitForAbortApplication);
+
+            await StartStreamAsync(1, _browserRequestHeaders, endStream: true);
+
+            Assert.Equal(Http2Connection.ConnectionState.Open, _connection.State);
+            Assert.Equal(1, _connection.HighestOpenedStreamId);
+
+            _connection.StopProcessingNextRequest();
+            Assert.True(_connection.StateTransitionEvents[Http2Connection.ConnectionState.Closing].Wait(5000));
+
+            VerifyGoAway(await ReceiveFrameAsync(), Int32.MaxValue, Http2ErrorCode.NO_ERROR);
+
+            await SendRstStreamAsync(1);
+
+            Assert.True(_connection.StateTransitionEvents[Http2Connection.ConnectionState.Closed].Wait(5000));
+            VerifyGoAway(await ReceiveFrameAsync(), 1, Http2ErrorCode.NO_ERROR);
+
+            await WaitForAllStreamsAsync();
+        }
+
+        [Fact]
+        public async Task AcceptNewStreamsDuringClosingConnection()
+        {
+            await InitializeConnectionAsync(_echoApplication);
+
+            await StartStreamAsync(1, _browserRequestHeaders, endStream: false);
+
+            Assert.Equal(Http2Connection.ConnectionState.Open, _connection.State);
+            Assert.Equal(1, _connection.HighestOpenedStreamId);
+
+            await SendGoAwayAsync();
```

Comment:
It still seems bizzar for the client to start new streams after sending a GoAway. "The GOAWAY frame (type=0x7) is used to initiate shutdown of a connection or to signal serious error conditions.". You didn't do anything specific to allow it, correct? You just didn't disallow it?

Reply:
The point here is to put the connection into a closing state. I did this via GOAWAY though I should have used StopProcessingNextRequest since that's more realistic. There's no functional difference though (other than an extra GOAWAY frame)